aead/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+//! A set of traits designed to support authenticated encryption.
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use core::{
+    convert::TryFrom,
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+};
+use failure::Fail;
+use generic_array::{
+    typenum::{Unsigned, U0},
+    ArrayLength,
+};
+use rand_core::{CryptoRng, RngCore};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialze};
+
+/// An enum describing possible failure modes
+#[cfg(feature = "serde")]
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Fail, Hash, Ord, PartialEq, PartialOrd, Serialize)]
+pub enum AeadError {
+    /// The nonce provided is the wrong size for this algorithm.
+    #[fail(display = "The nonce given is the wrong size for the algorithm")]
+    InvalidNonceSize,
+    /// The ciphertext buffer is not a multiple of the block size
+    #[fail(display = "The buffer given is the wrong size for the algorithm")]
+    InvalidBufferSize,
+    /// The "bytes used" provided is larger than the provided buffer
+    #[fail(display = "The buffer usage is larger than the buffer size")]
+    InvalidBufferUsed,
+    /// The authentication tag is not the correct size
+    #[fail(display = "The authentication tag is the wrong size for the algorithm")]
+    InvalidTagSize,
+    /// The cipher key was not the size of a block
+    #[fail(display = "The cipher key given on creation was the wrong size")]
+    InvalidCipherKeySize,
+    /// The initialization vector was not the size of a block
+    #[fail(display = "The IV given on creation was the wrong size")]
+    InvalidIvSize,
+    /// The authentication tag is not the size expected by this algorithm
+    #[fail(display = "The key used for the authentication algorithm was the wrong size")]
+    InvalidTagKeySize,
+
+    #[fail(
+        display = "The given ciphertext is larger than the maximum size allowed by this algorithm"
+    )]
+    CiphertextTooLarge,
+    #[fail(
+        display = "The given plaintext is larger than the maximum size allowed by this algorithm"
+    )]
+    PlaintextTooLarge,
+
+    /// The MAC failed to validate. This should be used to indicated a modified
+    /// ciphertext, modified additional data, or key mismatch.
+    #[fail(display = "The MAC failed to validate")]
+    MacFailure,
+    /// There was an error parsing the padding of this message
+    #[fail(display = "The padding was not correct")]
+    PaddingFailure,
+}
+
+/// An enum describing possible failure modes
+#[cfg(not(feature = "serde"))]
+#[derive(Clone, Copy, Debug, Eq, Fail, Hash, Ord, PartialEq, PartialOrd)]
+pub enum AeadError {
```

Comment:
I think it might be worth considering an opaque error type here, and in general look at some papers, resources, and case studies on misuse resistance in the design of AEAD APIs.

This sort of information, leaked to the attacker, can potentially be used for attacks. Whether or not that is possible will depend on the code which impl's this trait, but I think it's very much worth considering.

Reply:
Even beyond that, AEADs built using the Encode-then-Encipher paradigm (HCTR, HCH, AEZ, and various others), which provides the maximal misuse resistance (RAE, in Rogaway's terminology), are incapable of distinguishing many of these. As a result, there exist AEADs that will have to choose _which kind_ of misrepresentation they will perform, because this error type forces them to distinguish between things they cannot.